Microsoft.Toolkit/Diagnostics/Guard.Comparable.Generic.cs

```diff
@@ -0,0 +1,371 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsEqualTo<T>(T value, T target, string name)
+            where T : notnull, IEquatable<T>
+        {
+            if (!value.Equals(target))
+            {
+                ThrowHelper.ThrowArgumentExceptionForIsEqualTo(value, target, name);
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the input value must be not equal to a specified value.
+        /// </summary>
+        /// <typeparam name="T">The type of input values to compare.</typeparam>
+        /// <param name="value">The input <typeparamref name="T"/> value to test.</param>
+        /// <param name="target">The target <typeparamref name="T"/> value to test for.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="value"/> is == <paramref name="target"/>.</exception>
+        /// <remarks>The method is generic to avoid boxing the parameters, if they are value types.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsNotEqualTo<T>(T value, T target, string name)
+            where T : notnull, IEquatable<T>
+        {
+            if (value.Equals(target))
+            {
+                ThrowHelper.ThrowArgumentExceptionForIsNotEqualTo(value, target, name);
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the input value must be a bitwise match with a specified value.
+        /// </summary>
+        /// <typeparam name="T">The type of input values to compare.</typeparam>
+        /// <param name="value">The input <typeparamref name="T"/> value to test.</param>
+        /// <param name="target">The target <typeparamref name="T"/> value to test for.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="value"/> is not a bitwise match for <paramref name="target"/>.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsBitwiseEqualTo<T>(T value, T target, string name)
+            where T : unmanaged
+        {
+            /* Include some fast paths if the input type is of size 1, 2, 4 or 8.
+             * In those cases, just reinterpret the bytes as values of an integer type,
+             * and compare them directly, which is much faster than having explicitly
+             * loop through every single byte. This also allows for more expressive error
+             * messages, since the entire input values can be expressed as hexadecimal values.
+             * The conditional branches below are known at compile time by the JIT compiler,
+             * so that only the right one will actually be translated into native code. */
+            if (typeof(T) == typeof(byte) ||
```

Comment:
Since we're using the T4 stuff already, would it not be faster to have method overloads for the compiler to pick the right path combination vs. having ifs at runtime?

Reply:
It doesn't actually make a difference in this case, as those checks are only available in IL. Once the JIT compiles the method (or .NET Native on UWP), you'd get a different type-specialized method for each different type, just like if we had used a T4 template in this ase (yay reified generic types! 😄).

You can double check that the final asm is exactly the same, I've setup a sample file for you [here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8B6YgAgBkBLMGAO1xgBMyMJWALGMgOgDkAogBUyAMQgBXOk2wYqEOmSlMYUMgu7Qy+aN2wBzWDHz0MuHgFgAUKTJCuvQSPHLZ8xQBsa9Rrk5U/ADMqD242MgBPSSVpVU5uAFkASREvWgYYSxtyAGUYMIcKJIBhAT4cgTJg0LIqRQwHAAcoCAArGDAMMhaILsCtHVhaun6ofDcFLOsAAQAmAEYZgAYyafmeACUpORMeYoh8RpDVPKgAN29cAG5lshyObChGimxgTe2qExurGdmAZmsAG9rGRQas/qt5khVig7DBcBh5gAKHKNbB0AA8dQwAD4yLg0XQ0MMusAAJQgsHAqxg2lVXTYMAcMhI7FkbDDfGEik0ul8gDiEkeTB4SVwACEqBgAO4BGACACOQo8QggyOwxOAxLo2BMEECSOwZLJ31pAF9KaDLeDIdDprChPCMLMUYSsXRcVz0cS2eTrdS+aDRjBGczWR72ZyCeieYG6YLhaKJVLZYxFcrVShDZrtbqYPrDcbTWCLVZSzMIWtoWioHJsB5VrMyAmoEwgWDrQBtABSUv59FUNCRGAijXzBuxxoAul2EjAGhAmElDh4kXOF0uVwB5RpyBTmACC+kM8NwVDOMCSdC8dDq+jJM95oOmlahMLIYslMrl6frqvmGJCDiSIiGc9YSDAxIiBgjz6POxJrCsOomLGgbSlwQwiCAMTjDqcFtk+ZABoGVCBCyI5jgWQhkmQAC8tGsKO47hhgxrWrSxFxqCEhsmBHgQQAqleXQMQJDDYIEmQHrggHEjxHrAbAZF8RBJrsYG8ldDBUBwRgQkRqJ4mSTw0myUo2KKTAZHabpak/IRcakSyKkwPpXQAIQMTZ85uahXFEep/n2C00oHjpEgmB6AgILQu7uOIUDJt+aZKn+EBIi5xLeRguYocW/mlnGhWBjAHiMIFAUOYGSlkMAEQYNwLkbFZdHdC1Ym4BJUkyUImr1TAlnKeBMB2f5oI1XVDWsLB87NWRDE1R1XUmT1fUNYN006fOo1jb6/W4Ps2ytUtxk5FQABeMBboEgHAXZFW0qMLJslQrVLFctRkBitX7YdHofVQADUgN+VxnFjaCk2NcN4r9cdRlSUwTBIjVTVWT6O0Q5DcPZbDU2GZ1xkHkjKMtdlc0Y98D0kWRGUw3DnmbbpeMjdTfLg1jYLBRAoXhZFGDRbFe50AlSWpvKqUqulmVM/BZDISN+VY8VBUVSrZDq7O84cIuy6NKu6465u+s7sLh7HrAuBnheV43neD7Wi+trvp+KY/pLmYsWQMtstlCHzEheagxzYJOXT/HcIz2WgxxbN2BwIVhfoEWmILMBxQootfuLv5S+HEFZTNOXy0HSvmtaZq0taWsbnrBva7r24Zww/AQLbdT24+tJO1W77c7zyf82nzdZ27KUZhAt0gd7w1QbLxeISXKEVehqjcFhOHogYzD+hVDQhfLMDSmQScp1FMXp8LSIACQAEQAAqPHmDVqICCuV0igIUcx1FmjRATy16JGYAY9uDjAwEyYk0o8BfUBC5M0eI6gIhDCwfUsDsoINvrlRWFdrCliAA). You can see that `Test1`, using the generic method, and `Test2` (type-specific) produce the same final native code.

Also keep in mind that for types that can be bitwise-compared directly (eg. integer types), you can also use `Guard.IsEqualTo`, that'll work as well. The main difference is just in the error message, as `IsEqualTo` will display values in base 10, whereas `IsBitwiseEqualTo` will display them in hex format.

Another reason why I went with a single generic here was to try not to have another file with 1000 lines of code just for this single API 😅

Reply:
Yay for smart compilers! :)